engine/char.3
.\"	$Csoft: char.3,v 1.6 2002/02/05 05:49:43 vedge Exp $
.\"
.\" Copyright (c) 2001 CubeSoft Communications, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistribution of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistribution in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of CubeSoft Communications, nor the names of its
.\"    contributors may be used to endorse or promote products derived from
.\"    this software without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.\"	$OpenBSD: mdoc.template,v 1.6 2001/02/03 08:22:44 niklas Exp $
.\"
.Dd JANUARY 30, 2002
.Dt CHAR 3
.Os
.Sh NAME
.Nm char
.Nd AGAR character implementation
.Sh SYNOPSIS
.Fd #include <engine/char.h>
.Sh DESCRIPTION
The character structure represents a movable character occupying one or more
nodes on the map.
.Pp
.Bl -bullet -compact
.It
Current level and exp
.It
Current anx maximal HP/MP
.It
Status ailments (regen, poison, ...)
.It
Flags (focused, dash, ...)
.It
Directional/speed information
.El
.Sh ENGINE
.nr nS 1
.Ft "struct char *"
.Fn char_create "char *name" "char *descr" "int maxhp" "int maxmp" "int flags"
.Pp
.Ft "int"
.Fn char_link "void *obj"
.Pp
.Ft "void"
.Fn char_destroy "void *obj"
.Pp
.nr nS 0
.Pp
.Fn char_create
allocates a new character structure. 
Once it is in a consistent state, a call to
.Fn char_link
is made so the character can start handling time (and events, if the
character is focused).
A character is destroyed and unlinked from all maps with a call to
.Fn char_destroy .
.Fn char_event
delivers a SDL event to the character layer, it is called either
directly from the engine or from a particular character implementation.
Meaningful events include:
.Pp
.Bl -bullet -compact
.It
SDL_JOYAXISMOTION
.It
SDL_JOYBUTTONDOWN
.It
SDL_JOYBUTTONUP
.It
SDL_KEYDOWN
.It
SDL_KEYUP
.El
.Sh CHARACTER MOTION
.nr nS 1
.Ft "void"
.Fn char_setspeed "struct character *ch" "Uint32 speed"
.Pp
.Ft "void"
.Fn char_add "struct character *ch" "struct map *map" "int x" "int y" 
.Pp
.Ft "void"
.Fn char_canmove "struct character *ch" "int nx" "int ny"
.Pp
.Ft "void"
.Fn char_move "struct character *ch" "int nx" "int ny"
.Pp
.Ft "void"
.Fn char_del "struct character *ch" "struct map *map" "int x" "int y" 
.Pp
.Ft "void"
.Fn char_setsprite "struct character *ch" "int soffs"
.Pp
.Ft "void"
.Fn char_setanim "struct character *ch" "int aoffs"
.Pp
.nr nS 0
.Pp
.Fn char_setspeed
readjusts the character's timer.
The
.Fa speed
argument should be something between 0 and the character's maxspeed.
The display's refresh rate will be added to this interval.
.Pp
.Fn char_add
positions the character
.Fa ch
on
.Fa map
at the given coordinates, and
.Fn char_del
removes any reference to
.Fa ch
at the given coordinates.
.Fn char_canmove
returns 0 if the character is able to move to new coordinates
on the current map.
.Fn char_move
moves the character
.Fa ch
to new coordinates on the current map.
.Fn char_setsprite
sets the active sprite, and
.Fn char_setanim
sets the active animation.
.Sh SEE ALSO
.Xr agar 3


CHANGELOG.md
# Changelog
All notable changes to Agar will be documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.8.0] - Unreleased
### Added
- [**Style Editor**](https://libagar.org/man3/AG_StyleEditor): Display computed style rules in addition to per-instance overrides.
- [**AG_Button**](https://libagar.org/man3/AG_Button): New style attributes `text-align` and `vertical-align`.
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Introduce the `style()` method, called by the style compiler to parse widget-specific style attributes into an efficient binary representation.
- [**AG_Object**](https://libagar.org/man3/AG_Object): New function `AG_AccessVariable_NoLock()` (lockless variant of `AG_AccessVariable()`).
- [**AG_Object**](https://libagar.org/man3/AG_Object): Document the `AG_OBJECT_BOUND_EVENTS` flag.

### Removed
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove the "window-detached" event. It is redundant since the generic "detached" event does the same job.

### Changed
- Major performance improvements in the style compiler. Since it runs whenever widgets are resized, this should lead to better fluidity and overall lower CPU usage when Agar windows and container widgets are resized.
- [**AG_Object**](https://libagar.org/man3/AG_Object): In `AG_ObjectDestroy()`, do not call `AG_ObjectDetach()` on child objects. Instead, just cancel any active timers and null their parent pointers without raising a "detached" event. This allows us to perform garbage collection more efficiently.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): Make `AG_TLIST_NO_BGLINES` the default.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): In `AG_TlistNewPolledMs()`, accept a `ms` argument of 0 to disable periodical updates (the caller may still trigger updates with "tlist-poll").
- In `AG_WidgetCopyStyle()`, prevent redundant locking by using `AG_AccessVariable_NoLock()`.

### Fixed
- [**AG_Widget**](https://libagar.org/man3/AG_Widget): Fix redundant "detached" handler invocations. Improve overall efficiency of the garbage-collection step after windows are destroyed.
- [**AG_Console**](https://libagar.org/man3/AG_Console): Prevent a redundant lock operation in the backend of `AG_ConsoleMsg()`.

## [1.7.1] - 2025-11-15
### Added
- An awesome new Agar logo. Thanks [eightbit](https://8bitprodigy.github.io)!
- Added support for the [Cmake](https://cmake.org) build system.
- Added support for the [Meson](https://mesonbuild.com) build system. Thanks [coleman](https://git.sr.ht/~coleman/)!
- New functions [**AG_InitVideoSDL2**](https://libagar.org/man3/AG_InitVideoSDL2) and [**AG_SetVideoSurfaceSDL2**](https://libagar.org/man3/AG_SetVideoSurfaceSDL2) for integrating with an existing SDL2 display context. Thanks Brigham Keys!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Add "Any File" type.
- [**AG_Tlist**](https://libagar.org/man3/AG_Tlist): New function `AG_TlistCopy()`. Copy all items from a source to a destination `AG_Tlist`.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New constructor `AG_ComboNewFn()`. Sets a `combo-expanded` handler.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): New member `nVisItems`. Set the number of items to show by default in expansions.
- [**AG_Checkbox**](https://libagar.org/man3/AG_Checkbox): New functions `AG_CheckboxText()` and `AG_CheckboxTextS()` to update the text label.
- Tweak the default color scheme. Document style defaults as a comment block in `gui/style.css`.
- New Ada bindings for the `Window`, `Box`, `Button`, `Checkbox` and `Combo` widget classes. Updated `ada/gui/demo`.

### Removed
- [**AG_Db**](https://libagar.org/man3/AG_Db): Remove support for DB5.
- [**AG_Window**](https://libagar.org/man3/AG_Window): Remove an unnecessary counter variable in `SizeAllocate()`.

### Changed
- Small cosmetic changes to the default color scheme. Set default font size back to 14pt.
- [**AG_Titlebar**](https://libagar.org/man3/AG_Titlebar): Enable the `AG_BUTTON_EXCL` optimization on titlebar buttons.
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Allow the dropdown list to be populated by a `combo-expanded` handler.
- [**AG_UCombo**](https://libagar.org/man3/AG_UCombo): Allow the dropdown list to be populated by a `ucombo-expanded` handler.

### Fixed
- [**AG_Combo**](https://libagar.org/man3/AG_Combo): Make it again possible to statically initialize `list` before `combo-expanded`. Restores compatibility pre-1.6. Thanks Wally!
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix widget geometries not updating when switching to a different Type filter.
- [**AG_FileDlg**](https://libagar.org/man3/AG_FileDlg): Fix compilation where `GLOB_TILDE` is unavailable.
- SDL2 drivers: Require at least version 2.0.22 of SDL2 (`SDL_HINT_MOUSE_AUTO_CAPTURE` is necessary for best user experience).
- [**AG_ProgressBar**](https://libagar.org/man3/AG_ProgressBar): Make `padding` work as expected in progress bar. Thanks scaramacai!
- [**cocoa**](https://libagar.org/man3/AG_DriverCocoa): Fix a compilation error.
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Fix a keyboard handling error. Don't pass `KeyRelease` events to `XwcLookupString()`. Fixes issue #74. Thanks @tsesani and @scaramacai!
- [**glx**](https://libagar.org/man3/AG_DriverGLX): Introduce the `no-im` (disable input methods) driver option.
- [**MAP**](https://libagar.org/man3/MAP): Fix a crash when attempting to create an object if `pLibsRoot` is NULL.
- [**AG_Web**](https://libagar.org/man3/AG_Web): Fix inline compilation problem due to a typo in `net/web.h`.
- On Darwin platforms, build with `_DARWIN_C_SOURCE` globally. Fixes issue #77. Thanks @kwhr0!
- On FreeBSD check for availability of `CLOCK_SECOND` in addition to `HAVE_CLOCK_GETTIME`.
- Test for the `dirfd()` function at compilation.

trunk/engine/char.c
/*	$Csoft: char.c,v 1.10 2002/02/05 05:49:03 vedge Exp $	*/

/*
 * Copyright (c) 2001 CubeSoft Communications, Inc.
 * <http://www.csoft.org>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistribution of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistribution in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of CubeSoft Communications, nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <pthread.h>
#include <glib.h>
#include <SDL.h>

#include <engine/engine.h>

#define JOY_UP		0x01
#define JOY_DOWN	0x02
#define JOY_LEFT	0x04
#define JOY_RIGHT	0x08

static Uint32	char_time(Uint32, void *);
static void	char_event(struct character *, SDL_Event *);
static void	char_joybutton(struct character *, SDL_Event *);
static void	char_key(struct character *, SDL_Event *);

struct character *curchar;

struct character *
char_create(char *name, char *desc, int maxhp, int maxmp, int flags)
{
	struct character *ch;
	
	ch = (struct character *)malloc(sizeof(struct character));
	if (ch == NULL) {
		return (NULL);
	}

	object_create(&ch->obj, name, desc, flags);
	ch->event_hook = char_event;
	ch->map = NULL;
	ch->x = -1;
	ch->y = -1;

	ch->flags = 0;
	ch->curspeed = 1;
	ch->maxspeed = 50;
	ch->level = 0;
	ch->exp = 0;
	ch->age = 0;
	ch->maxhp = maxhp;
	ch->maxmp = maxmp;
	ch->hp = maxhp;
	ch->mp = maxmp;
	ch->seed = lrand48();
	ch->effect = 0;

	char_setanim(ch, 0);	 /* XXX */

	dprintf("%s: hp %d/%d mp %d/%d seed 0x%lx\n",
	    ch->obj.name, ch->hp, ch->maxhp, ch->mp, ch->maxmp, ch->seed);

	return (ch);
}

int
char_link(void *ob)
{
	struct character *ch = (struct character *)ob;
	
	if (pthread_mutex_lock(&world->lock) == 0) {
		SLIST_INSERT_HEAD(&world->wcharsh, ch, wchars);
		pthread_mutex_unlock(&world->lock);
	} else {
		perror("world");
		return (-1);
	}

	if (object_link(ob) < 0) {
		return (-1);
	}
	
	ch->timer = SDL_AddTimer(60 + (ch->maxspeed - ch->curspeed),
	    char_time, ch);
	if (ch->timer == NULL) {
		fatal("SDL_AddTimer: %s\n", SDL_GetError());
		return (-1);
	}

	return (0);
}

void
char_destroy(struct object *ob)
{
	struct character *ch = (struct character *)ob;

	SDL_RemoveTimer(ch->timer);

	if (ch->flags & CHAR_FOCUS) {
		char_unfocus(ch);
	}

	if (pthread_mutex_lock(&ch->map->lock) == 0) {
		if (ch->map != NULL) {
			MAP_DELREF(ch->map, ch->x, ch->y, ob, -1);
		}
		pthread_mutex_unlock(&ch->map->lock);
	} else {
		perror(ch->map->obj.name);
	}

	if (pthread_mutex_lock(&world->lock) == 0) {
		SLIST_REMOVE(&world->wcharsh, ch, character, wchars);
		pthread_mutex_unlock(&world->lock);
	} else {
		perror(world->obj.name);
	}
}

/* See if ch can move to nx:ny in its map. */
int
char_canmove(struct character *ch, int nx, int ny)
{
	struct node *me;

	me = &ch->map->map[nx][ny];

	return ((me->flags & NODE_WALK) ? 0 : -1);
}

static void
char_event(struct character *ch, SDL_Event *ev)
{
	switch (ev->type) {
	case SDL_JOYBUTTONDOWN:
	case SDL_JOYBUTTONUP:
		char_joybutton(ch, ev);
		break;
	case SDL_KEYDOWN:
	case SDL_KEYUP:
		char_key(ch, ev);
		break;
	}
}

/* Change current animation. */
int
char_setsprite(struct character *ch, int soffs)
{
	ch->flags &= ~(CHAR_ANIM);
	ch->curoffs = soffs;

	return (0);
}

/* Let the player control this character. */
int
char_focus(struct character *ch)
{
	ch->flags |= CHAR_FOCUS;
	curchar = ch;

	return (0);
}

int
char_unfocus(struct character *ch)
{
	ch->flags &= CHAR_FOCUS;
	curchar = NULL;
	
	return (0);
}

/* Change current sprite. */
int
char_setanim(struct character *ch, int aoffs)
{
	ch->flags |= CHAR_ANIM;
	ch->curoffs = aoffs;

	return (0);
}

/* Position the character on m:x,y. */
int
char_add(struct character *ch, struct map *m, int x, int y)
{
	if (ch->flags & CHAR_ANIM) {
		MAP_ADDANIM(m, x, y, (struct object *)ch, ch->curoffs);
	} else {
		MAP_ADDSPRITE(m, x, y, (struct object *)ch, ch->curoffs);
	}
	
	ch->map = m;
	ch->x = x;
	ch->y = y;

	return (0);
}

/* Delete any reference to the character on m:x,y. */
int
char_del(struct character *ch, struct map *m, int x, int y)
{
	MAP_DELREF(m, x, y, (struct object *)ch, -1);

	/* Map reference (m:x,y) is now undefined. */

	return (0);
}

/* Move the character from its current position to nx,ny. */
int
char_move(struct character *ch, int nx, int ny)
{
	char_del(ch, ch->map, ch->x, ch->y);
	char_add(ch, ch->map, nx, ny);

	ch->x = nx;
	ch->y = ny;

	return (0);
}

void
char_setspeed(struct character *ch, Uint32 speed)
{
	ch->curspeed = speed;

	if (ch->curspeed >= ch->maxspeed) {
		ch->curspeed = ch->maxspeed;
	} else if (ch->curspeed <= 0) {
		ch->curspeed = 1;
	}
	if (SDL_RemoveTimer(ch->timer)) {
		ch->timer = SDL_AddTimer(
		    ch->map->view->fps + (ch->maxspeed - ch->curspeed),
		    char_time, ch);
	}
}

static Uint32
char_time(Uint32 ival, void *obp)
{
	struct object *ob = (struct object *)obp;
	struct character *ch = (struct character *)ob;
	struct map_aref *aref;

	/* XXX limit one sprite/anim. */
	aref = node_arefobj(&ch->map->map[ch->x][ch->y], ob, -1);
#ifdef DEBUG
	if (aref == NULL) {
		dprintf("%s is not at %dx%d\n", ob->name, ch->x, ch->y);
	}
#endif

	/*
	 * Vertical soft scroll.
	 */
	if (aref->yoffs < 0) {
		ch->map->redraw++;
		dprintf("up move\n");
		if (aref->yoffs == -1) {	/* Once */
			char_setanim(ch, 1);
			if (char_canmove(ch, ch->x, ch->y - 1) < 0) {
				dprintf("blocked!\n");
				aref->yoffs = 0;
				goto xoffsck;
			}
		}
		aref->yoffs -= 5;
		dprintf("yoffs is now %d\n", aref->yoffs);
		if (aref->yoffs <= -ch->map->view->tileh) {
			char_move(ch, ch->x, ch->y - 1);
			aref->yoffs = 0;
		}
	} else if (aref->yoffs > 0) {
		dprintf("down move\n");
		if (aref->yoffs == 1) {
			if (char_canmove(ch, ch->x, ch->y + 1) < 0) {
				dprintf("blocked!\n");
				aref->yoffs = 0;
				goto xoffsck;
			}
			char_setanim(ch, 2);
		}
		aref->yoffs++;
		if (aref->yoffs >= ch->map->view->tileh) {
			char_move(ch, ch->x, ch->y + 1);
			aref->yoffs = 0;
		}
	}

xoffsck:
	/*
	 * Horizontal soft scroll.
	 */
	if (aref->xoffs < 0) {
		dprintf("left move\n");
		if (aref->xoffs == -1) {
			if (char_canmove(ch, ch->x - 1, ch->y) < 0) {
				dprintf("blocked!\n");
				aref->xoffs = 0;
				goto xoffsck;
			}
			char_setanim(ch, 3);
		}
		aref->xoffs--;
		if (aref->xoffs <= -ch->map->view->tilew) {
			char_move(ch, ch->x - 1, ch->y);
			aref->xoffs = 0;
		}
	} else if (aref->xoffs > 0) {
		dprintf("right move\n");
		if (aref->xoffs == 1) {
			if (char_canmove(ch, ch->x + 1, ch->y) < 0) {
				dprintf("blocked!\n");
				aref->xoffs = 0;
				goto ailmentck;
			}
			char_setanim(ch, 4);
		}
		aref->xoffs++;
		if (aref->xoffs >= ch->map->view->tilew) {
			char_move(ch, ch->x + 1, ch->y);
			aref->xoffs = 0;
		}
	}

ailmentck:
	/* Assume various status ailments. */

#if 0
	if (nme->flags & NODE_BIO) {
		decrease(&ch->hp, 1, 1);
		dprintf("bio. hp = %d/%d\n", ch->hp, ch->maxhp);
	} else if (nme->flags & NODE_REGEN) {
		increase(&ch->hp, 1, ch->maxhp);
		dprintf("regen. hp = %d/%d\n", ch->hp, ch->maxhp);
	}

	if (nme->flags & NODE_SLOW) {
		/* XXX rate */
		nme->v1 = -10;
		char_setspeed(ch, ch->curspeed + nme->v1);
		dprintf("slow. speed = %d\n", ch->curspeed);
	} else if (nme->flags & NODE_HASTE) {
		/* XXX rate */
		nme->v1 = 10;
		char_setspeed(ch, ch->curspeed + nme->v1);
		dprintf("haste. speed = %d\n", ch->curspeed);
	}
#endif

	return (ival);
}

static void
char_joybutton(struct character *ch, SDL_Event *ev)
{
	static SDL_Event nev;
	
	dprintf("key %d\n", ev->jbutton.button);
	
	nev.type = (ev->type == SDL_JOYBUTTONUP) ?
	    SDL_KEYUP : SDL_KEYDOWN;
	
	/* XXX remap */
	switch (ev->jbutton.button) {
	case 2:
		nev.key.keysym.sym = SDLK_d;	/* Dash */
		break;
	/* ... */
	}
	SDL_PushEvent(&nev);
}

static void
char_key(struct character *ch, SDL_Event *ev)
{
	int set;
	struct map_aref *aref;

	set = (ev->type == SDL_KEYDOWN);
	/* XXX limit one sprite/anim. */
	aref = node_arefobj(&ch->map->map[ch->x][ch->y],
	    (struct object *)ch, -1);

	switch (ev->key.keysym.sym) {
	case SDLK_d:	/* Dash */
		if (ev->type == SDL_KEYDOWN) {
			ch->flags |= CHAR_DASH;
			char_setspeed(ch, 40);	/* XXX increment */
		} else if (ev->type == SDL_KEYUP) {
			ch->flags &= ~(CHAR_DASH);
			char_setspeed(ch, 1);	/* XXX restore speed */
		}
		break;
	case SDLK_UP:
		aref->yoffs = set ? -1 : 0;
		break;
	case SDLK_DOWN:
		aref->yoffs = set ? 1 : 0;
		break;
	case SDLK_LEFT:
		aref->xoffs = set ? -1 : 0;
		break;
	case SDLK_RIGHT:
		aref->xoffs = set ? 1 : 0;
		break;
	default:
		break;
	}
}

#ifdef DEBUG

void
char_dump(struct character *ch)
{
	printf("%3d. %10s lvl %d (ex %.2f) hp %d/%d mp %d/%d at %s:%dx%d\n",
	    ch->obj.id, ch->obj.name, ch->level, ch->exp, ch->hp, ch->maxhp,
	    ch->mp, ch->maxmp, ch->map->obj.name, ch->x, ch->y);
	printf("\t\t< ");
	if (ch->flags & CHAR_FOCUS)
		printf("focused ");
	printf(">");
}

#endif /* DEBUG */

